src/main/java/io/r2dbc/postgresql/client/ExtendedQueryMessageFlow.java

```diff
@@ -69,15 +76,55 @@ private ExtendedQueryMessageFlow() {
      * @return the messages received in response to the exchange
      * @throws IllegalArgumentException if {@code bindings}, {@code client}, {@code portalNameSupplier}, or {@code statementName} is {@code null}
      */
-    public static Flux<BackendMessage> execute(Publisher<Binding> bindings, Client client, PortalNameSupplier portalNameSupplier, String statementName, String query, boolean forceBinary) {
-        Assert.requireNonNull(bindings, "bindings must not be null");
+    public static Flux<BackendMessage> execute(Binding binding, Client client, PortalNameSupplier portalNameSupplier, String statementName, String query, boolean forceBinary,
+                                               int fetchSize) {
+        Assert.requireNonNull(binding, "binding must not be null");
         Assert.requireNonNull(client, "client must not be null");
         Assert.requireNonNull(portalNameSupplier, "portalNameSupplier must not be null");
         Assert.requireNonNull(statementName, "statementName must not be null");
-
-        return client.exchange(Flux.from(bindings)
-            .flatMap(binding -> toBindFlow(binding, portalNameSupplier, statementName, query, forceBinary))
-            .concatWith(Mono.just(Sync.INSTANCE)));
+        if (fetchSize == NO_LIMIT) {
+            return Flux.defer(() -> {
+                String portal = portalNameSupplier.get();
+                Flux<FrontendMessage> bindFlow = toBindFlow(binding, portal, statementName, query, forceBinary, fetchSize).concatWithValues(new Close(portal, PORTAL))
+                    .concatWithValues(Sync.INSTANCE);
+                return client.exchange(bindFlow)
+                    .as(Operators::discardOnCancel)
+                    .doOnDiscard(ReferenceCounted.class, ReferenceCountUtil::release);
+            });
+        }
+        return Flux.defer(() -> {
+            String portal = portalNameSupplier.get();
+            Flux<FrontendMessage> bindFlow = toBindFlow(binding, portal, statementName, query, forceBinary, fetchSize).concatWithValues(Flush.INSTANCE);
+            DirectProcessor<FrontendMessage> requestsProcessor = DirectProcessor.create();
+            FluxSink<FrontendMessage> requestsSink = requestsProcessor.sink();
+            AtomicBoolean isCanceled = new AtomicBoolean(false);
+            return client.exchange(bindFlow.concatWith(requestsProcessor))
+                .handle((BackendMessage message, SynchronousSink<BackendMessage> sink) -> {
```

Comment:
Another thought crossed my mind: Right now, a query is required to fully complete before another one can be executed.

 What if we would break down cursored queries into smaller conversations that terminate either with `CommandComplete` or `PortalSuspended` messages? 

That way, we could interleave different queries and we would have smaller conversations in total. We could also factor in the current demand before requesting more data. I'm not sure how long a single portal can be kept open though.

Reply:
Error handling will be really tricky if we're going to allow multiple queries at one time. After server sends ErrorResponse it discards every request until it find Sync (that's why we're sending Sync after error). Also Sync will commit transaction (if there were now explicit transactions) and this will close every portal. We can have our connection in some really weird state after that: neverending conversations (since backend discarded requests) or closed portals (since backend closed them on Sync) are possible.

Reply:
Okay, I see that interleaving doesn’t make sense. I’m going to merge and polish the PR in the next days and take a look on fetch size with unparametrized statements.

Reply:
I was going to implement unparametrized statements fetch size in this PR somewhere near weekend, but ok, you can do it if you want to.